Feature/332 kms key auth

thijshberg opened the pull request:
Adds the KMS feature for the SSH capability

Includes some tests and a deployment example

It currently works as an interface to an underlying directory with keys. There is are API endpoints for listing all keys, adding a key, and revoking a key (which moves and renames the underlying files). It is activated and configured by environment variables.

If the feature is not configured at startup there is no method to enable it later. The routes are also not instantiated if the feature is disabled.

thijshberg committed:
Add Key Management Store to the SSH capability

This is done via a .env flag and directory indication.
We use the kms property of the user_auth field in CACAO for the
reference to the key.

The key identifier is just the name of the file in the directory as given in
.env. This is convenient for persistence and interaction with other
tasks. The keys are cached. There is room for an api interacting with
the system.

There is also some extra information in the ssh logging, which helps
with identifying errors in the ssh layer.

thijshberg committed:
Add Key Management Store to the SSH capability

This is done via a .env flag and directory indication.
We use the kms property of the user_auth field in CACAO for the
reference to the key.

The key identifier is just the name of the file in the directory as given in
.env. This is convenient for persistence and interaction with other
tasks. The keys are cached. There is room for an api interacting with
the system.

There is also some extra information in the ssh logging, which helps
with identifying errors in the ssh layer.

thijshberg committed:
Added example for the ssh KMS

There is now an example playbook using the KMS system.

There is also a testing docker config. This does not use the same setup
as the existing ssh example unfortunately.

The example requires some work to set up the KMS folder. You should
first set up an example keypair in the deployment example, and then set
up a folder for the KMS system containing those keys (either copying or
using the deployment location).

thijshberg committed:
Added example for the ssh KMS

There is now an example playbook using the KMS system.

There is also a testing docker config. This does not use the same setup
as the existing ssh example unfortunately.

The example requires some work to set up the KMS folder. You should
first set up an example keypair in the deployment example, and then set
up a folder for the KMS system containing those keys (either copying or
using the deployment location).

thijshberg committed:
Added REST API for SSH key management

This makes it possible to get the list of keys, add keys manually, and
revoke keys. Revoking does not delete keys but just moves and renames
the underlying files to prevent irrevertible mishaps.

This should serve as a starting point for consumers such as
soarca-gui.

thijshberg committed:
Added REST API for SSH key management

This makes it possible to get the list of keys, add keys manually, and
revoke keys. Revoking does not delete keys but just moves and renames
the underlying files to prevent irrevertible mishaps.

This should serve as a starting point for consumers such as
soarca-gui.

thijshberg committed:
Added key management unit tests

This also removes some bugs from the existing code

thijshberg committed:
Added key management unit tests

This also removes some bugs from the existing code

thijshberg committed:
Improve example deployment

thijshberg committed:
Improve example deployment

thijshberg committed:
Added helper script for KMS example setup

thijshberg committed:
Added helper script for KMS example setup

thijshberg committed:
Updated linting errors

thijshberg committed:
Updated linting errors

thijshberg committed:
Removed a public-private mixup

thijshberg committed:
Removed a public-private mixup

thijshberg committed:
Removed lint error

thijshberg committed:
Removed lint error

MaartendeKruijf commented on pkg/core/capability/ssh/ssh.go:
```diff
@@ -115,13 +131,13 @@ func getSession(config ssh.ClientConfig, target cacao.AgentTarget) (*ssh.Session
 	host := CombinePortAndAddress(target.Address, target.Port)
 	client, err := ssh.Dial("tcp", host, &config)
 	if err != nil {
-		log.Error(err)
```
Errorf won't be supported by other log framworks so we only use Error

MaartendeKruijf replied:
Also the error from the Dial is self explanatory so no need to change

MaartendeKruijf commented on pkg/core/capability/ssh/ssh.go:
```diff
@@ -173,7 +192,7 @@ func close(client *ssh.Client) {
 		err := client.Close()
 		if err != nil {
-			log.Error(err)
+			log.Errorf("Closing: %s", err)
```
Idem

MaartendeKruijf commented on pkg/core/capability/ssh/ssh.go:
```diff
@@ -115,13 +131,13 @@ func getSession(config ssh.ClientConfig, target cacao.AgentTarget) (*ssh.Session
 	session, err := client.NewSession()
 	if err != nil {
-		log.Error(err)
+		log.Errorf("Session: %s", err)
```
Idem

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+package ssh
```
Key Management should  be its own module as it's not part of the capability, it should be injected on execution. Now SSH is not testable without key management working or partly initialised.

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+	}
+	if n_read < len(public) {
+		return fmt.Errorf("write error: did not write whole public key file")
+	}
```
This check is already performed in the write library

```
if n != len(b) {
		err = io.ErrShortWrite
	}

```

MaartendeKruijf commented on pkg/api/api.go:
```diff
@@ -7,10 +7,12 @@ import (
 	reporter_handler "soarca/pkg/api/reporter"
 	status_handler "soarca/pkg/api/status"
 	"soarca/pkg/core/capability/manual/interaction"
+	"soarca/pkg/core/capability/ssh"
```
capability should not be included here the only exception is manual due to its inherent external nature.

MaartendeKruijf commented on deployments/docker/testing/ssh-kms-test/prepare_keys.sh:
```diff
@@ -0,0 +1,6 @@
+mkdir -p ssh-keystore/
+ssh-keygen -q -N "" -f ssh-keystore/test
+mkdir -p config/.ssh/
+cat ssh-keystore/test.pub > config/.ssh/authorized_keys
```
Move to test pipeline as this test will never run

MaartendeKruijf commented on deployments/docker/testing/ssh-kms-test/docker-compose.yml:
```diff
@@ -0,0 +1,17 @@
+services:
+  openssh-server:
+    image: lscr.io/linuxserver/openssh-server:latest
```
Just use the plain ubuntu or alpine, this introduces new dependencies for testing

MaartendeKruijf commented on internal/controller/controller.go:
```diff
@@ -269,6 +269,17 @@ func initializeCore(app *gin.Engine) error {
 		return err
 	}
 
+	if utils.GetEnv("ENABLE_SSH_KMS", "false") == "true" {
```
I think KMS should always be enabled, you can just pre load it with keys if you want by providing the path.

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+	if err != nil {
+		return nil, err
+	}
+	passphrase := utils.GetEnv("SSH_KMS_PASSPHRASE", "")
```
Downstream modules should never interact with the environment

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+	}
+	passphrase := utils.GetEnv("SSH_KMS_PASSPHRASE", "")
+	file_buffer := make([]byte, 2048)
+	if _, err := file.Read(file_buffer); err != nil {
```
use  `contents, err := io.ReadAll(file)` for getting the full file without making (a potentially to small) buffer. Maybe we should load the files from a different module and inject them via a secure keystore.

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+				return fmt.Errorf("public key (%s) parsing error: %s", filename, err)
+			}
+			management.cached_keys[private_filename] = KeyPair{Public: public, Private: private}
+		}
```
This could be a function

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+func (handler *KeyManagementHandler) GetKeys(context *gin.Context) {
+	keyInfo := handler.Manager.ListAllNames()
```
No logging

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+func (handler *KeyManagementHandler) Refresh(context *gin.Context) {
+	err := handler.Manager.Refresh()
```
No logging

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+		Manager: manager,
+	}
+}
+
```
Api misses Swagger config

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+	})
+}
+
+func SendErrorResponse(context *gin.Context, status int, message string, orginal_call string) {
```
this is already in api.error

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+	keyname := context.Param("keyname")
+	jsonData, err := io.ReadAll(context.Request.Body)
+	if err != nil {
+		log.Trace("Submit key has failed: ", err.Error())
```
This is an error so it should log as one

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+	}
+	var key api.KeyManagementKey
+	if err := json.Unmarshal(jsonData, &key); err != nil {
+		log.Trace("Submit key failed to unmarshal: ", err.Error())
```
idem

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+		return
+	}
+	if err := handler.Manager.Insert(key.Public, key.Private, keyname); err != nil {
+		log.Trace("Submit key failed to insert: ", err.Error())
```
idem

MaartendeKruijf commented on pkg/api/keymanagement/keymanagement_api.go:
```diff
@@ -0,0 +1,93 @@
+		return
+	}
+	context.JSON(http.StatusOK, gin.H{
+		"status":  200,
```
This should be a model not raw JSON

MaartendeKruijf commented on pkg/core/capability/ssh/ssh.go:
```diff
@@ -69,18 +70,19 @@ func executeCommand(session *ssh.Session,
+	if sessionErr != nil && sessionErr.Error() != "EOF" {
+		// The ssh api is subtle, and it can happen that we get EOF as an error.
+		// This is likely not an error, as the session can also be closed by the host.
+		log.Errorf("Close: %s", sessionErr)
```
In any case log the EOF

MaartendeKruijf commented on pkg/core/capability/ssh/ssh.go:
```diff
@@ -156,8 +172,11 @@ func CheckSshAuthenticationInfo(authentication cacao.AuthenticationInformation)
 
 	switch authentication.Type {
 	case "user-auth":
-		if strings.TrimSpace(authentication.Password) == "" {
-			return errors.New("password is empty")
+		if strings.TrimSpace(authentication.Password) == "" && !authentication.Kms {
+			return errors.New("password is empty and KMS is not indicated")
+		}
+		if authentication.Kms && strings.TrimSpace(authentication.KmsKeyIdentifier) == "" {
+			return errors.New("KMS is indicated but no identifier is given")
```
This should be different, first check for KMS, because when its set username and password may not be populated https://docs.oasis-open.org/cacao/security-playbooks/v2.0/cs01/security-playbooks-v2.0-cs01.html#_Toc152256508 we could choose to just ignore when set but that is technically out of spec

MaartendeKruijf commented on test/unittest/mocks/mock_utils/test-key.pub:
```diff
@@ -0,0 +1 @@
+ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGzijTr3feGJz8CuxKjR+z1xROTQAdza843VmpLrxV/u thijs@PC-44318
```
Maybe don't leave PII here

MaartendeKruijf commented on deployments/docker/testing/ssh-kms-test/prepare_keys.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+set -u
+mkdir -p ssh-keystore/
+ssh-keygen -q -N "" -f ssh-keystore/test
```
Dynamic keys could lead to testing errors I think we should generate one upfront, we can move it to a docker file

MaartendeKruijf commented on examples/ssh-kms-playbook.json:
```diff
@@ -0,0 +1,117 @@
+        "example"
+    ],
+    "authentication_info_definitions": {
+        "user-auth--b7ddc2ea-9f6a-4e82-8eaa-be202e942090": {
+            "type": "user-auth",
+            "kms":true,
+            "kms_key_identifier": "test",
+            "username": "linuxserver.io"
```
This should be non standard

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+	cached_keys    map[string]KeyPair
+}
+
+var globalKeyManagement *KeyManagement
```
Manager should keep global status

MaartendeKruijf commented on pkg/core/capability/ssh/keymanagement.go:
```diff
@@ -0,0 +1,252 @@
+	file_buffer := make([]byte, 2048)
+	if _, err = file.Read(file_buffer); err != nil {
+		return nil, err
+	}
```
[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

thijshberg committed:
Fixed the logger level configuration

The Severity.fromString method did not do anything with the base it was
called from. This meant that all logs took the global level instead of
the per-package option.

thijshberg committed:
Fixed the logger level configuration

The Severity.fromString method did not do anything with the base it was
called from. This meant that all logs took the global level instead of
the per-package option.

thijshberg committed:
Changed KMS to use database backend

thijshberg committed:
Changed KMS to use database backend

thijshberg committed:
Updated ssh-kms deployment and added ci

thijshberg committed:
Updated ssh-kms deployment and added ci

thijshberg committed:
Updated ci build to properly fetch keys

thijshberg committed:
Updated ci build to properly fetch keys

thijshberg committed:
Smaller fixes from review

thijshberg committed:
Smaller fixes from review

thijshberg committed:
Fixed name clash

thijshberg committed:
Fixed name clash

thijshberg committed:
More capitalization errors

thijshberg committed:
More capitalization errors

thijshberg committed:
Fixing lint errors

thijshberg committed:
Fixing lint errors

thijshberg committed:
Removed cached keys

thijshberg committed:
Removed cached keys

thijshberg commented:
The folder-based structure has been removed, the backing is now the same as the playbook api. That is, a mongo option and a simple in-memory option.

This removes the need for a cache in the keymanagement component. The keymanagement component is now more of a front-end for dealing with key parsing and hiding the indirection of the database.

MaartendeKruijf committed:
Revert "Fixed the logger level configuration"

This reverts commit 23315e1a7bb6ae7bd80c2f34c6acf09022a58ed7.

MaartendeKruijf committed:
Reverted: In memory database changes

MaartendeKruijf committed:
Moved to temporary package

MaartendeKruijf committed:
Reverted Mongo changes

MaartendeKruijf committed:
renamed key management

MaartendeKruijf committed:
Allowed any creation for collection in mongo anticipating refactor

MaartendeKruijf committed:
Ad-Hoc collection creation

MaartendeKruijf committed:
Updated stray go.mod

MaartendeKruijf committed:
Refactored key management test

MaartendeKruijf committed:
Refactored KMS test

MaartendeKruijf committed:
Moved test-keys keymanagement to test/kms

MaartendeKruijf committed:
Initialisation cleanups

MaartendeKruijf committed:
Updated docs api spec

MaartendeKruijf committed:
Fixed a typo in the path

MaartendeKruijf committed:
Fixed error if KMS is unset that SOARCA would crash

MaartendeKruijf committed:
Added the ENABLE_SSH_KMS flag into the logic

MaartendeKruijf committed:
Introduced KMS interfaces

MaartendeKruijf committed:
Changed ssh capability dependency to IKms

MaartendeKruijf committed:
Changed keymanagement api to IKeyManagementApi

MaartendeKruijf committed:
Updated api description

MaartendeKruijf committed:
Added info log to keymanagement

hidde-jan commented:
Quick question @thijshberg, does this create (or use) an internally managed KMS? I think integration with existing KMS (gcloud, hashicorp, etc) is much more useful and keeps the hard part (securely managing credentials) away from soarca.

thijshberg commented:
> Quick question @thijshberg, does this create (or use) an internally managed KMS? I think integration with existing KMS (gcloud, hashicorp, etc) is much more useful and keeps the hard part (securely managing credentials) away from soarca.

Essentially yes. The current version would serve as a starting point for that, as it leaves the actual implementation to be generic.

hidde-jan commented:
I can see an in-memory KMS make sense from a local development and testing POV 👍 

However, I think soarca should probably not be used to actually **manage** key material. Ideally it shouldn't need to have permissions to create, update or revoke credentials in an (external) KMS, just request (temporary) credentials it needs to perform steps.

What would make sense in my mind is a read-only in-memory and simple file based KMS for development and testing. This wouldn't need API endpoints at all.
